### ISSUE TITLE
Gradients do not propagate when iterating over dataset in graph mode

### ISSUE THREAD

**Issue by terhorst** [tensorflow/tensorflow#36306]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS X 10.15
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2.0-dev20200128
- Python version: 3.8

**Describe the current behavior**
Gradients become `None` if iterating over a `tf.data.Dataset` in graph mode. The correct answer is returned in eager mode.
**Describe the expected behavior**
Gradients in eager and graph mode should be the same.

**Code to reproduce the issue**
 Example:

```
data = tf.range(10, dtype=tf.float32)
ds = tf.data.Dataset.from_tensor_slices(data)
u = tf.Variable(1., dtype=tf.float32)

with tf.GradientTape(persistent=True) as g:
    loss1 = ds.reduce(tf.constant(0.), lambda x, y: x + (y - u) ** 2)
    loss2 = tf.reduce_sum([(y - u) ** 2 for y in ds.as_numpy_iterator()])
print(g.gradient(loss1, u))  # returns None
print(g.gradient(loss2, u))  # returns correct gradient
```
The same behavior is observed when using `for x in ds: ...` inside of a `@tf.function`.

Gist [here](https://colab.research.google.com/gist/terhorst/b51638f5e9657eb7f44553922c5a93c5/untitled3.ipynb).

**Comment by terhorst** [comment]
A workaround is to move the GradientTape inside of the accumulated function, and accumulate gradients alongside the loss.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.2.0-dev20200129 and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/1641e618441ed823560057b4d24c6d75/untitled604.ipynb).

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.3-rc1, 2.4.0-dev20200721 and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/750d4f1c916c795e98adad09c28bf03c/untitled163.ipynb).Thanks!

**Comment by saikumarchalla** [comment]
@terhorst tf.data does not support gradients. Closing this issue as of now since it is the intended behavior.Thanks!
